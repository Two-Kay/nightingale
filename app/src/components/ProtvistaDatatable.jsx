import React, { useEffect, useState } from "react";
import ProtvistaDatatable from "protvista-datatable";
import ProtvistaTrack from "protvista-track";
import ProtvistaManager from "protvista-manager";
import ProtvistaNavigation from "protvista-navigation";
import { transformData } from "protvista-feature-adapter";
import { load } from "data-loader";
import loadWebComponent from "../utils/load-web-component";
import readmeContent from "../../../packages/protvista-datatable/README.md";
import Readme from "./Readme";

// const columnConfig = {
//   consequence: {
//     label: "Evidences",
//     child: true,
//     resolver: (d) => {
//     },
//   },
//   ftId: {
//     label: "Feature ID",
//     child: true,
//     resolver: (d) => d.ftId,
//   },
// };
const ProtvistaDatatableWrapper = () => {
  const [data, setData] = useState();

  loadWebComponent("protvista-datatable", ProtvistaDatatable);
  loadWebComponent("protvista-manager", ProtvistaManager);
  loadWebComponent("protvista-track", ProtvistaTrack);
  loadWebComponent("protvista-navigation", ProtvistaNavigation);

  useEffect(() => {
    async function fetchData() {
      const loadedData = await load(
        "https://www.ebi.ac.uk/proteins/api/features/P05067?categories=MOLECULE_PROCESSING"
      );
      const transformedData = transformData(loadedData?.payload);
      setData(transformedData);
    }
    fetchData();
  }, []);

  // console.log(data);

  return (
    <>
      <Readme content={readmeContent} />
      <h2>Track with data-loader</h2>
      <protvista-manager attributes="length displaystart displayend variantfilters highlight selectedid">
        {/* <protvista-navigation length="770" />
        <protvista-track
          id="my-protvista-track"
          length="770"
          layout="non-overlapping"
        ></protvista-track> */}
        <protvista-datatable>
          <table>
            <thead>
              <tr>
                <th data-filter="ft_key">Feature key</th>
                <th>Description</th>
                <th>Positions</th>
              </tr>
            </thead>
            <tbody>
              {data?.map((row, i) => (
                <tr
                  key={i}
                  data-start={row.start}
                  data-end={row.end}
                  data-id={`${row.start}-${row.end}`}
                >
                  <td data-filter="ft_key">{row.type}</td>
                  <td>{row.description}</td>
                  <td>
                    {row.start}-{row.end}
                  </td>
                  {/* <td data-collapsed>{row.type}</td> */}
                </tr>
              ))}
            </tbody>
          </table>
        </protvista-datatable>
      </protvista-manager>
    </>
  );
};

export default ProtvistaDatatableWrapper;
