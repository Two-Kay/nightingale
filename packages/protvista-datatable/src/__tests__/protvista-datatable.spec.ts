import { fireEvent, getByRole, getByText, waitFor } from "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";

import ProtvistaDatatable from "../protvista-datatable";

describe("protvista-datatable", () => {
  beforeAll(() => {
    // TODO remove when the definition is part of the import
    window.customElements.define("protvista-datatable", ProtvistaDatatable);
  });

  afterEach(() => {
    document.querySelector("protvista-datatable").remove();
    document.documentElement.innerHTML = "";
  });

  test("it should display the datatable correctly", async () => {
    document.documentElement.innerHTML = `
<protvista-datatable>
    <table>
        <thead>
            <tr>
                <th>Col A</th>
                <th>Col B</th>
                <th>Col C</th>
                <th>Col D</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>Row 1 A</td>
                <td>Row 1 B</td>
                <td>Row 1 C</td>
                <td>Row 1 D</td>
            </tr>
            <tr>
                <td>Row 2 A</td>
                <td>Row 2 B</td>
                <td>Row 2 C</td>
                <td>Row 2 D</td>
            </tr>
        </tbody>
</protvista-datatable>
    `;
    const rendered = document.querySelector("protvista-datatable");
    expect(rendered).toMatchSnapshot();
  });

  it("should handle grouped rows", async () => {
    document.documentElement.innerHTML = `
    <protvista-datatable>
        <table>
            <thead>
                <tr>
                    <th>Col A</th>
                    <th>Col B</th>
                    <th>Col C</th>
                </tr>
            </thead>
            <tbody>
                <tr data-id="id1">
                    <td>Row 1 A</td>
                    <td>Row 1 B</td>
                    <td>Row 1 C</td>
                    <td>Row 1 D</td>
                </tr>
                <tr data-group-for="id1">
                    <td>Long row</td>
                </tr>
            </tbody>
    </protvista-datatable>
        `;
    const rendered = document.querySelector<HTMLElement>("protvista-datatable");
    const lr = getByText(rendered, "Long row") as HTMLTableCellElement;
    expect(lr).not.toBeVisible();
    expect(lr.colSpan).toBe(4);
    const button = getByRole(rendered, "button");
    fireEvent.click(button);
    await waitFor(() => {
      expect(lr).toBeVisible();
    });
  });
});
