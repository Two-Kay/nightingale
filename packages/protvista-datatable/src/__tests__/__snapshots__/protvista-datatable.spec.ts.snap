// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protvista-datatable it should display the datatable correctly 1`] = `
<protvista-datatable
  class="feature"
>
  
    
  <table>
    
        
    <thead>
      
            
      <tr>
        <th
          class="pd-group-column-header"
        />
        
                
        <th>
          Col A
        </th>
        
                
        <th>
          Col B
        </th>
        
                
        <th>
          Col C
        </th>
        
                
        <th>
          Col D
        </th>
        
            
      </tr>
      
        
    </thead>
    
        
    <tbody>
      
            
      <tr
        class="odd"
      >
        
                
        <td
          class="pd-group-trigger"
        />
        <td>
          Row 1 A
        </td>
        
                
        <td>
          Row 1 B
        </td>
        
                
        <td>
          Row 1 C
        </td>
        
                
        <td>
          Row 1 D
        </td>
        
            
      </tr>
      
            
      <tr
        class="even"
      >
        
                
        <td
          class="pd-group-trigger"
        />
        <td>
          Row 2 A
        </td>
        
                
        <td>
          Row 2 B
        </td>
        
                
        <td>
          Row 2 C
        </td>
        
                
        <td>
          Row 2 D
        </td>
        
            
      </tr>
      
        
    </tbody>
    

    
  </table>
</protvista-datatable>
`;
